### PR TITLE
Add `match_phrase`  to `FilterQueryBuilder` (part of #185)

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
* Added `MatchPhraseQuery` which inherits `LuceneQuery` and works as `MatchQuery`
* Added `MATCH_PHRASE` function registration
* Added tests `MATCH_PHRASE`

Signed-off-by: Yury Fridlyand [yuryf@bitquilltech.com](mailto:yuryf@bitquilltech.com)

### CODE REVIEW

**Comment by MaxKsyunz** [comment]
@Yury-Fridlyand , the Java CI action fails.  Have a look at this run: https://github.com/Bit-Quill/opensearch-project-sql/runs/6218062469?check_suite_focus=true
and this one: https://github.com/Bit-Quill/opensearch-project-sql/runs/6217945785?check_suite_focus=true
(these SQL Java CI / build (17) (push) and (pull_request)

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import org.opensearch.sql.opensearch.storage.script.filter.lucene.LuceneQuery;
+
+public class MatchPhraseQuery extends LuceneQuery {
+  private final BiFunction<MatchPhraseQueryBuilder, ExprValue, MatchPhraseQueryBuilder> analyzer =
+          (b, v) -> b.analyzer(v.stringValue());
+  private final BiFunction<MatchQueryBuilder, ExprValue, MatchQueryBuilder> slop =
+          (b, v) -> b.maxExpansions(Integer.parseInt(v.stringValue()));
+  private final BiFunction<MatchPhraseQueryBuilder, ExprValue, MatchPhraseQueryBuilder> zeroTermsQuery =
+          (b, v) -> b.zeroTermsQuery(
+                  org.opensearch.index.search.MatchQuery.ZeroTermsQuery.valueOf(v.stringValue()));
```
Comment (MaxKsyunz) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java:27]: org.opensearch.index ... should be imported at the start of the file.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  ImmutableMap<Object, Object> argAction = ImmutableMap.builder()
+          .put("analyzer", analyzer)
+          .put("slop", slop)
+          .put("zero_terms_query", zeroTermsQuery)
+          .build();
+
+  @Override
+  public QueryBuilder build(FunctionExpression func) {
+    Iterator<Expression> iterator = func.getArguments().iterator();
+    NamedArgumentExpression field = (NamedArgumentExpression) iterator.next();
```
Comment (MaxKsyunz) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java:38]: Having type on both sides looks redundant. This [document](https://openjdk.java.net/projects/amber/guides/lvti-style-guide#G3) agrees,  but checkStyle might be set up differently.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    NamedArgumentExpression query = (NamedArgumentExpression) iterator.next();
```
Comment (MaxKsyunz) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java:39]: Another good candidate for var.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    MatchPhraseQueryBuilder queryBuilder = QueryBuilders.matchPhraseQuery(
+            field.getValue().valueOf(null).stringValue(),
+            query.getValue().valueOf(null).stringValue());
+    while (iterator.hasNext()) {
+      NamedArgumentExpression arg = (NamedArgumentExpression) iterator.next();
```
Comment (MaxKsyunz) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java:44]: Looks like a good place to use var.

**Review by MaxKsyunz — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/opensearch/sql/expression/DSL.java [hunk 5]
```diff
@@ -654,4 +654,9 @@ public FunctionExpression match(Expression... args) {
     return (FunctionExpression) repository
         .compile(BuiltinFunctionName.MATCH.getName(), Arrays.asList(args.clone()));
   }
+
+  public FunctionExpression match_phrase(Expression... args) {
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/expression/DSL.java:658]: Should be camelCase here?

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java (hunk 1 above)
Comment (Yury-Fridlyand) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java:27]: This block of code copied from `opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchQuery.java` as is. Perhaps, we need to do code clean up in scope of another task.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java (hunk 2 above)
Comment (Yury-Fridlyand) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java:38]: This block of code copied from `opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchQuery.java` as is. Perhaps, we need to do code clean up in scope of another task.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java (hunk 3 above)
Comment (Yury-Fridlyand) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java:39]: This block of code copied from `opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchQuery.java` as is. Perhaps, we need to do code clean up in scope of another task.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java (hunk 4 above)
Comment (Yury-Fridlyand) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java:44]: This block of code copied from `opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchQuery.java` as is. Perhaps, we need to do code clean up in scope of another task.

File: core/src/main/java/org/opensearch/sql/expression/DSL.java (hunk 5 above)
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/DSL.java:658]: This block of code copied from `MatchQuery` implementation as is. Perhaps, we need to do code clean up in scope of another task.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.opensearch.storage.script.filter.lucene.relevance;
+
+import com.google.common.collect.ImmutableMap;
+import java.util.Iterator;
+import java.util.function.BiFunction;
+
+import org.opensearch.index.query.*;
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java]: Thank you, fixed in 628d9afa.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java (hunk 2 above)
Comment (MaxKsyunz) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java:39]: Sounds good.

**Review by MaxKsyunz — APPROVED** [review]

**Pull request merged**